src/newrelic/processor.go

```diff
@@ -6,819 +6,812 @@
+        client:              p.cfg.Client,
 		RequestHeadersMap:   app.connectReply.RequestHeadersMap,
-		// Splitting large payloads is limited to applications that have
-		// distributed tracing on. That restriction is a saftey measure
-		// to not overload the backend by sending two payloads instead
-		// of one every 60 seconds.
-		splitLargePayloads: app.info.Settings["newrelic.distributed_tracing_enabled"] == true,
-		blocking:           ph.Blocking,
-	}
+        // Splitting large payloads is limited to applications that have
+        // distributed tracing on. That restriction is a saftey measure
+        // to not overload the backend by sending two payloads instead
+        // of one every 60 seconds.
+        splitLargePayloads: app.info.Settings["newrelic.distributed_tracing_enabled"] == true,
+        blocking:           ph.Blocking,
+    }
 
-	harvestByType(ph.AppHarvest, &args, harvestType)
+    harvestByType(ph.AppHarvest, &args, harvestType)
 }
 
 func (p *Processor) processHarvestError(d HarvestError) {
-	h, ok := p.harvests[d.id]
-	if !ok {
-		// Very possible:  One harvest goroutine may encounter a ErrForceRestart
-		// before this.
-		log.Debugf("unable to process harvest response %q for unknown id %q",
-			d.Reply, d.id)
-		return
-	}
-
-	app := h.App
-	log.Warnf("app %q with run id %q received %s", app, d.id, d.Err)
-
-	switch {
-	case collector.IsDisconnect(d.Err):
-		app.state = AppStateDisconnected
-		p.shutdownAppHarvest(d.id)
-	case collector.IsLicenseException(d.Err):
-		// I think this is unlikely to ever happen (the invalid license
-		// exception should trigger during the connect), but it is included
-		// here for defensiveness.
-		app.state = AppStateInvalidLicense
-		p.shutdownAppHarvest(d.id)
-	case collector.IsRestartException(d.Err):
-		app.state = AppStateUnknown
-		p.shutdownAppHarvest(d.id)
-		p.considerConnect(app)
-	case (d.Err == collector.ErrPayloadTooLarge) ||
-		(d.Err == collector.ErrUnsupportedMedia):
-		// Do not call the failed harvest fn, since we do not want to save
-		// the data.
-	default:
-		d.data.FailedHarvest(h.Harvest)
-	}
+    h, ok := p.harvests[d.id]
+    if !ok {
+        // Very possible:  One harvest goroutine may encounter a ErrForceRestart
+        // before this.
+        log.Debugf("unable to process harvest response %q for unknown id %q",
+        d.Reply, d.id)
+        return
+    }
+
+    app := h.App
+    log.Warnf("app %q with run id %q received %s", app, d.id, d.Reply.Err)
+
+    if d.Reply.ShouldSaveHarvestData() {
+        d.data.FailedHarvest(h.Harvest)
+    }
+    switch {
+        case d.Reply.IsDisconnect() || app.state == AppStateDisconnected:
+            app.state = AppStateDisconnected
+            p.shutdownAppHarvest(d.id)
+        case d.Reply.IsRestartException() || app.state == AppStateRestart:
+            app.state = AppStateUnknown
+            p.shutdownAppHarvest(d.id)
+            p.considerConnect(app)
+    }
```

Comment:
Don't we still need the default case, just in case?

Reply:
The logic has changed slightly such that the default should no longer be necessary. At the top of the block, if the response code indicates saving data, we do that before any of the restart/disconnect logic. So at this point, calling failedHarvest() is no longer mutually exclusive with the cases in this switch block.

In the case where we receive an unknown response code, we will not save the harvest data, we will not disconnect, and we will not restart. It will simply discard the harvest and continue on.

Reply:
Maybe at least an error message if we hit the default saying we received unknown and are discarding?

Reply:
That's a little tricky, but I can add it. Basically, it cannot be the default case (because there are currently non-successful codes that neither direct for a restart or shutdown), but I can add a check against an entire list of known codes.